Add option to specify mDNS advertised IP address for HomeKit Bridge

mback2k opened the pull request:
## Breaking Change:

<!-- What is breaking and why we have to break it. Remove this section only if it was NOT a breaking change. -->
This is not a breaking change, as the default behaviour will be unchanged.

## Description:

This makes use of HAP-python's new feature in version 2.6.0
that allows to specify the mDNS advertised IP address.

This is a requirement for the following use cases:
- Running Home Assistant behind a NAT, e.g. inside Docker.
- Running it on a system with multiple interfaces there
  the default IP address, DNS entry and hostname diverge.

The forwarding of the required mDNS packets can be done with
an avahi-daemon based gateway, e.g. by using enable-reflector=yes.

Whitespace changes were performed due to black and flake8.

**Related PRs:**
Makes use of the following addition to HAP-python: https://github.com/ikalchev/HAP-python/pull/203
which was merged into HA via: https://github.com/home-assistant/home-assistant/pull/26783
with commit: https://github.com/home-assistant/home-assistant/commit/ed21019b7a4c746b443b7d1192db11880da06c90

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io):** home-assistant/home-assistant.io#10410

## Example entry for `configuration.yaml` (if applicable):
Example configuration for HA running inside Docker on a machine with IP address `192.168.1.100`:
```yaml
homekit:
  name: "Test HA"
  ip_address: "0.0.0.0"
  advertise_ip: "192.168.1.100"
  port: 51827
  filter:
    include_domains:
      - climate
      - light
      - sensor
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:

>   - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
>   - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
>   - [x] Untested files have been added to `.coveragerc`.

**Above is not relevant, as tests were updated. Manifest and requirements are already in place.**

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

mback2k committed:
Add options to specify advertised IP and MAC for HomeKit Bridge

This makes use of HAP-python's new feature in version 2.6.0
that allows to specify the mDNS advertised IP and MAC address.

This is a requirement for the following use cases:
- Running Home Assistant behind a NAT, e.g. inside Docker.
- Running it on a system with multiple interfaces there
  the default IP address, DNS entry and hostname diverge.

The forwarding of the required mDNS packets can be done with
an avahi-daemon based gateway, e.g. by using enable-reflector=yes.

Specifying the MAC address makes it possible to identify an
accessory in case HA is run inside a ephemeral docker container.

Whitespace changes were performed due to black and flake8.

homeassistant commented:
Hi @mback2k, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#26791).

Once you do that we will be able to review and accept this pull request.

Thanks!

mback2k committed:
Update tests for HomeKit Bridge due to IP and MAC advertising

Whitespace changes were performed due to black and flake8.

mback2k commented:
I am still working on the tests and will mark this PR ready for merge once they are ready.

mback2k commented:
All local and CI tests pass. All items in checklist completed. This PR is ready for review.

mback2k commented:
@adrum @fuzzie360 @Jc2k since I saw you being involved in other issues/PRs regarding the HomeKit integration: what do you think about this PR and what needs to be done in order to get it merged? I think this PR is quite small, straight forward and since it does not brake backwards compatibility it should be pretty easy to merge.

Jc2k commented:
I maintain homekit_controller so can only offer my generic HomeKit protocol knowledge and generic Home Assistant knowledge, i'm not too familiar with how the homekit part works.

Can you just confirm if what this calls a mac address is actually a Device ID? Device ID looks like a mac adress and is how a device is uniquely identified in HomeKit land, but is actually nothing to do with L2 ethernet. A Device ID is supposed to be random, persist across reboots and regenerated "on factory reset". It also looks like its used when authenticating a secure session (https://github.com/ikalchev/HAP-python/blob/b4d2f47a99a5767ffa4b1528085b9b8293ca08da/pyhap/hap_server.py#L409) so changing it after pairing might break the pairing?

I am fine with the ip advertise part of this change, though.

mback2k commented:
Yeah, the MAC address does not need to be the real network interface phyiscal address. It is like you describe just a unique Device ID that needs to be static across reboots. That is why I am adding the option, because inside a Docker container the Device ID is currently MAC address based and that changes on every container restart.

I could add that information to the documentation and/or change the parameter "mac" to "device_id" or something similar.

Jc2k commented:
Personally my preference would be for it to be random and stored in the homekit state file. It looks like that is what is meant to happen in pyHAP (there is some code used to persist it in [here](https://github.com/ikalchev/HAP-python/blob/5e8a981edb2ce4379647297f3fd5e23e4088a01b/pyhap/encoder.py#L14)) - i don't know if that's broken somehow in HASS. That way it should just work for everyone without needing to configure anything extra.

mback2k commented:
I think your proposal to have it automatically generated and stored in the state files makes sense as the default behaviour, but I would still like to be able to overwrite it manually just in case.

At the moment the MAC address in the State is generated here:
https://github.com/ikalchev/HAP-python/blob/e845ad93503146974b342449dc0b78c3171a244a/pyhap/state.py#L21
by using:
https://github.com/ikalchev/HAP-python/blob/e845ad93503146974b342449dc0b78c3171a244a/pyhap/util.py#L54-L62

So you are right. It is already randomly generated. And it seems like the State of the HomeKit Bridge is already persisted including that generated MAC address:
https://github.com/ikalchev/HAP-python/blob/e845ad93503146974b342449dc0b78c3171a244a/pyhap/accessory_driver.py#L483-L491

So to me it looks like your proposal is already implemented. I just didn't figure it out during my initial tests that the MAC address is actually random instead of the "random" MAC address of the ephemeral Docker container.

mback2k commented:
@balloob Hey Paulus, do you mind me bringing this relatively small PR to your attention?

balloob commented:
Let's stop referring to it as a mac address (it's a device ID). We should not allow the user to set it. It's not their concern.

mback2k commented:
Ok, fine with me. I will remove the MAC address / device ID part in this PR and just keep the possibility to set the advertised IP address.

mback2k committed:
Remove the possibility to set the MAC address of the HomeKit Bridge

Since the MAC address is a random device ID, there is no need
for the user to be able to set a custom MAC address value for it.

Whitespace changes were performed due to black and flake8.

mback2k commented:
@balloob @Jc2k Okay, I think I took care of your concerns. Anything else that needs to be done to get this merged?

Jc2k commented:
Fine from my point of view :+1:

mback2k commented:
@balloob Thank you very much! :+1:
